docs/Advanced-Usage/Heterogeneous-SoCs.rst

```diff
@@ -0,0 +1,127 @@
+-------------------------------------------
+
+Adding a Hwacha accelerator is as easy as adding the ``DefaultHwachaConfig`` so that it can setup the Hwacha parameters and add itself to the ``BuildRoCC`` parameter.
+An example of adding a Hwacha to all tiles in the system is below.
+
+.. code-block:: scala
+
+    class DualBoomAndRocketWithHwachasConfig extends Config(
+      new WithNormalBoomRocketTop ++
+      new WithBootROM ++
+      new hwacha.DefaultHwachaConfig ++
+      new boom.system.WithRenumberHarts ++
+      new boom.common.WithRVC ++
+      new boom.common.DefaultBoomConfig ++
+      new boom.system.WithNBoomCores(2) ++
+      new freechips.rocketchip.subsystem.WithoutTLMonitors ++
+      new freechips.rocketchip.subsystem.WithNBigCores(1) ++
+      new freechips.rocketchip.system.BaseConfig)
+
+In this example, Hwachas are added to both BOOM tiles and to the Rocket tile.
+All with the same Hwacha parameters.
+
+Assigning Accelerators to Specific Tiles with MultiRoCC
+-------------------------------------------------------
+
+Located in ``generators/example/src/main/scala/ConfigMixins.scala`` is a mixin that provides support for adding RoCC accelerators to specific tiles in your SoC.
+Named ``MultiRoCCKey``, this key allows you to attach RoCC accelerators based on the ``hartId`` of the tile.
+For example, using this allows you to create a 8 tile system with a RoCC accelerator on only a subset of the tiles.
+An example is shown below with two BOOM cores, and one Rocket tile with a Hwacha attached.
+
+.. code-block:: scala
+
+    class DualBoomAndOneHwachaRocketConfig extends Config(
+      new WithNormalBoomRocketTop ++
+      new WithBootROM ++
+      new WithMultiRoCC ++
+      new WithMultiRoCCHwacha(0) ++ // put Hwacha just on hart0 which was renumbered to Rocket
+      new boom.system.WithRenumberHarts(rocketFirst = true) ++
+      new hwacha.DefaultHwachaConfig ++
+      new boom.common.WithRVC ++
+      new boom.common.DefaultBoomConfig ++
+      new boom.system.WithNBoomCores(2) ++
+      new freechips.rocketchip.subsystem.WithoutTLMonitors ++
+      new freechips.rocketchip.subsystem.WithNBigCores(1) ++
+      new freechips.rocketchip.system.BaseConfig)
+
+In this example, the ``WithRenumberHarts`` relabels the ``hartId``'s of all the BOOM/Rocket cores.
+Then after that is applied to the parameters, the ``WithMultiRoCCHwacha(0)`` is used to assign to ``hartId`` zero a Hwacha (in this case ``hartId`` zero is Rocket).
+Finally, the ``WithMultiRoCC`` mixin is called.
+This mixin sets the ``BuildRoCC`` key to use the ``MultiRoCCKey`` instead of the default.
+This must be used after all the RoCC parameters are set because it needs to override the ``BuildRoCC`` parameter.
+If this is used earlier in the configuration sequence, then MultiRoCC does not work.
+
+This mixin can be changed to put more accelerators on more cores by changing the arguments to cover more ``hartId``'s (i.e. ``WithMultiRoCCHwacha(0,1,3,6,...)``).
```

Comment:
Should we mention a the generator approach here if we don't want to hard-code a hartId? (use length or number of tiles to apply the accelerator only to the last tile, or something of that sort)

Reply:
If I understand you correctly, to do your proposal you need a bunch of extra mixins (WithLastAccel, WithBeginningAccel, etc). IMHO, this is a cleaner approach because it makes sure that the user knows what accel. is attached to what system.